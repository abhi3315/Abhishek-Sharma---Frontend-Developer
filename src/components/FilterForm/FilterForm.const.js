export const capsuleStatus = [
	{
		value: '',
		label: 'All',
	},
	{
		value: 'active',
		label: 'Active',
	},
	{
		value: 'destroyed',
		label: 'Destroyed',
	},
	{
		value: 'retired',
		label: 'Retired',
	},
	{
		value: 'unknown',
		label: 'Unknown',
	},
];

export const capsuleType = [
	{
		value: '',
		label: 'All',
	},
	{
		value: 'Dragon 1.0',
		label: 'Dragon 1.0',
	},
	{
		value: 'Dragon 1.1',
		label: 'Dragon 1.1',
	},
	{
		value: 'Dragon 2.0',
		label: 'Dragon 2.0',
	},
];
