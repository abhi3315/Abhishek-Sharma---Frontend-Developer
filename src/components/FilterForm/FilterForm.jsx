import { useFormik } from 'formik';

import { useCapsuleFilter } from '../../contexts/capsule-filter';
import { capsuleStatus, capsuleType } from './FilterForm.const';

function FilterForm() {
	const { setFilter } = useCapsuleFilter();

	const { handleSubmit, values, handleChange, resetForm } = useFormik({
		initialValues: {
			capsuleStatus: '',
			capsuleType: '',
			serialId: '',
		},
		onSubmit: (submittedData) => setFilter(submittedData),
		onReset: (initialValues) => setFilter(initialValues),
	});

	return (
		<div className='px-10 xl:px-32 mb-5 sm:mb-20'>
			<h2 className='dark:text-white text-4xl md:text-5xl font-bold text-center my-5 sm:my-20'>
				SpaceX Capsules
			</h2>
			<div className='block rounded-lg bg-white p-6 shadow-lg dark:bg-zinc-800 w-full'>
				<form className='grid grid-cols-1 gap-4 md:grid-cols-4' onSubmit={handleSubmit}>
					<div>
						<label
							htmlFor='capsule-status'
							className='block mb-2 text-sm font-medium text-gray-900 dark:text-white'
						>
							Capsule Status
						</label>
						<select
							name='capsuleStatus'
							id='capsule-status'
							data-testid='capsule-status'
							value={values.capsuleStatus}
							onChange={handleChange}
							className='bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500'
						>
							{capsuleStatus.map(({ value, label }) => (
								<option key={value} value={value}>
									{label}
								</option>
							))}
						</select>
					</div>
					<div>
						<label
							htmlFor='capsule-type'
							className='block mb-2 text-sm font-medium text-gray-900 dark:text-white'
						>
							Capsule Type
						</label>
						<select
							name='capsuleType'
							id='capsule-type'
							data-testid='capsule-type'
							value={values.capsuleType}
							onChange={handleChange}
							className='bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500'
						>
							{capsuleType.map(({ value, label }) => (
								<option key={value} value={value}>
									{label}
								</option>
							))}
						</select>
					</div>
					<div>
						<label
							htmlFor='serial-id'
							className='block mb-2 text-sm font-medium text-gray-900 dark:text-white'
						>
							Serial ID
						</label>
						<input
							type='text'
							id='serial-id'
							name='serialId'
							data-testid='serial-id'
							value={values.serialId}
							onChange={handleChange}
							className='bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500'
						/>
					</div>
					<div className='flex items-end justify-center'>
						<button
							type='submit'
							data-testid='submit-filters'
							className='rounded px-8 py-2 shadow-md font-medium uppercase text-white bg-green-600 hover:bg-green-800 transition-all duration-200 ease-linear'
						>
							Filter
						</button>
					</div>
				</form>
			</div>
			<div className='flex justify-end py-2 px-3'>
				<button
					type='button'
					onClick={resetForm}
					data-testid='clear-filters'
					className='text-red-500 dark:text-red-400 hover:text-red-600 dark:hover:text-red-600 flex transition-colors'
				>
					Clear Filters
					<svg
						xmlns='http://www.w3.org/2000/svg'
						fill='none'
						viewBox='0 0 24 24'
						strokeWidth={1.5}
						stroke='currentColor'
						className='w-6 h-6'
					>
						<path strokeLinecap='round' strokeLinejoin='round' d='M6 18L18 6M6 6l12 12' />
					</svg>
				</button>
			</div>
		</div>
	);
}

export default FilterForm;
