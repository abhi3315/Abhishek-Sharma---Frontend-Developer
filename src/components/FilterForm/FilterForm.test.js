import { act, cleanup, fireEvent, render } from '@testing-library/react';

import FilterForm from './FilterForm';

afterEach(cleanup);

describe('FilterForm', () => {
	it('should render correctly', () => {
		const { container } = render(<FilterForm />);
		expect(container).toMatchSnapshot();
	});

	it('should have intial form value to be empty', () => {
		const { getByTestId } = render(<FilterForm />);

		const capsuleStatusSelect = getByTestId('capsule-status');
		const capsuleTypeSelect = getByTestId('capsule-type');
		const capsuleSerialInput = getByTestId('serial-id');

		expect(capsuleStatusSelect.value).toBe('');
		expect(capsuleTypeSelect.value).toBe('');
		expect(capsuleSerialInput.value).toBe('');
	});

	it('should have form value to be updated', async () => {
		const { getByTestId } = render(<FilterForm />);

		const capsuleStatusSelect = getByTestId('capsule-status');
		const capsuleTypeSelect = getByTestId('capsule-type');
		const capsuleSerialInput = getByTestId('serial-id');

		await act(async () => {
			fireEvent.change(capsuleStatusSelect, { target: { value: 'active' } });
			fireEvent.change(capsuleTypeSelect, { target: { value: 'Dragon 1.0' } });
			fireEvent.change(capsuleSerialInput, { target: { value: 'C101' } });
		});

		expect(capsuleStatusSelect.value).toBe('active');
		expect(capsuleTypeSelect.value).toBe('Dragon 1.0');
		expect(capsuleSerialInput.value).toBe('C101');
	});

	it('should have form value to be updated and reset', async () => {
		const { getByTestId } = render(<FilterForm />);

		const capsuleStatusSelect = getByTestId('capsule-status');
		const capsuleTypeSelect = getByTestId('capsule-type');
		const capsuleSerialInput = getByTestId('serial-id');
		const resetButton = getByTestId('clear-filters');

		await act(async () => {
			fireEvent.change(capsuleStatusSelect, { target: { value: 'active' } });
			fireEvent.change(capsuleTypeSelect, { target: { value: 'Dragon 1.0' } });
			fireEvent.change(capsuleSerialInput, { target: { value: 'C101' } });
			fireEvent.click(resetButton);
		});

		expect(capsuleStatusSelect.value).toBe('');
		expect(capsuleTypeSelect.value).toBe('');
		expect(capsuleSerialInput.value).toBe('');
	});
});
