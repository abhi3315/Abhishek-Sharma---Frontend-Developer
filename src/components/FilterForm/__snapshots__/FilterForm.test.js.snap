// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterForm should render correctly 1`] = `
<div>
  <div
    class="px-10 xl:px-32 mb-5 sm:mb-20"
  >
    <h2
      class="dark:text-white text-4xl md:text-5xl font-bold text-center my-5 sm:my-20"
    >
      SpaceX Capsules
    </h2>
    <div
      class="block rounded-lg bg-white p-6 shadow-lg dark:bg-zinc-800 w-full"
    >
      <form
        class="grid grid-cols-1 gap-4 md:grid-cols-4"
      >
        <div>
          <label
            class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
            for="capsule-status"
          >
            Capsule Status
          </label>
          <select
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
            data-testid="capsule-status"
            id="capsule-status"
            name="capsuleStatus"
          >
            <option
              value=""
            >
              All
            </option>
            <option
              value="active"
            >
              Active
            </option>
            <option
              value="destroyed"
            >
              Destroyed
            </option>
            <option
              value="retired"
            >
              Retired
            </option>
            <option
              value="unknown"
            >
              Unknown
            </option>
          </select>
        </div>
        <div>
          <label
            class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
            for="capsule-type"
          >
            Capsule Type
          </label>
          <select
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
            data-testid="capsule-type"
            id="capsule-type"
            name="capsuleType"
          >
            <option
              value=""
            >
              All
            </option>
            <option
              value="Dragon 1.0"
            >
              Dragon 1.0
            </option>
            <option
              value="Dragon 1.1"
            >
              Dragon 1.1
            </option>
            <option
              value="Dragon 2.0"
            >
              Dragon 2.0
            </option>
          </select>
        </div>
        <div>
          <label
            class="block mb-2 text-sm font-medium text-gray-900 dark:text-white"
            for="serial-id"
          >
            Serial ID
          </label>
          <input
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
            data-testid="serial-id"
            id="serial-id"
            name="serialId"
            type="text"
            value=""
          />
        </div>
        <div
          class="flex items-end justify-center"
        >
          <button
            class="rounded px-8 py-2 shadow-md font-medium uppercase text-white bg-green-600 hover:bg-green-800 transition-all duration-200 ease-linear"
            data-testid="submit-filters"
            type="submit"
          >
            Filter
          </button>
        </div>
      </form>
    </div>
    <div
      class="flex justify-end py-2 px-3"
    >
      <button
        class="text-red-500 dark:text-red-400 hover:text-red-600 dark:hover:text-red-600 flex transition-colors"
        data-testid="clear-filters"
        type="button"
      >
        Clear Filters
        <svg
          class="w-6 h-6"
          fill="none"
          stroke="currentColor"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 18L18 6M6 6l12 12"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
